File: 24. Swap Nodes in Pairs/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head

        # weave in and out to extract nodes
        ptr = head
        dummy = ListNode(None)
        prev = dummy
        while ptr and ptr.next:
            tmp = ptr.next.next
            prev.next = ptr.next
            prev.next.next = ptr
            ptr.next = tmp
            prev = ptr
            ptr = tmp
        
        return dummy.next
                
        

File: 24. Swap Nodes in Pairs/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;
        ListNode node = head;
        dummy.next = head;
        ListNode tmp;
        
        while (node != null && node.next != null) {
            tmp = node.next;
            node.next = node.next.next;
            tmp.next = node;
            prev.next = tmp;
            prev = node;
            node = node.next;
        }
        return dummy.next;
    }
}
